abba/json/src/Parse.cpp

```diff
@@ -0,0 +1,480 @@
+	std::cout << "getWithoutWhiteSpace function called" << std::endl;
+	char	c = ' ';
+	while (c == ' ' || c == '\t' || c == '\n' || \
+			c == '\v' || c == '\f' || c == '\r')
+	{
+		_file->get(c);
+		if (!_file)
+			return c;
+	}
+	return c;
+}
+
+Parse::Type Parse::getToken()
+{
+	char		c;
+	Type	type;
+
+//	std::cout << "getToken function called" << std::endl;
+	if (_file->eof())
+	{
+		std::cout << "finish reading" << std::endl;
+	}
+	_prevPos = _file->tellg();
+	c = getWithoutWhiteSpace();
+	if (c == '"')
+	{
+		type.type = STRING;
+		_file->get(c);
+		while (c != '\n' && c != '"')
+		{
+			type.value += c;
+			_file->get(c);
+		}
+	}
+	else if (c == '[')
+	{
+//		std::cout << "array open" << std::endl;
+		type.type = ARRAY_OPEN;
+	}
+	else if (c == ']')
+	{
+//		std::cout << "array close" << std::endl;
+		type.type = ARRAY_CLOSE;
+	}
+	else if (c == '{')
+	{
+//		std::cout << "curly open" << std::endl;
+		type.type = CURLY_OPEN;
+	}
+	else if (c == '}')
+		type.type = CURLY_CLOSE;
+	else if (c == ':')
+		type.type = COLON;
+	else if (c == ',')
+		type.type = COMMA;
+	else if (c >= '0' && c <= '9')
+	{
+		type.type = NUMBER;
+		while (c >= '0' && c <= '9')
+		{
+			_prevPos = _file->tellg();
+			type.value += c;
+			_file->get(c);
+		}
+		_file->seekg(_prevPos);
+	}
+	else if (c == 'n')
+	{
+		type.type = NULL_TYPE;
+		while (isalpha(c))
+		{
+			_prevPos = _file->tellg();
+			type.value += c;
+			_file->get(c);
+		}
+		_file->seekg(_prevPos);
+	}
+	else if (c == 't' || c == 'f')
+	{
+		type.type = BOOLEAN;
+		while (isalpha(c))
+		{
+			_prevPos = _file->tellg();
+			type.value += c;
+			_file->get(c);
+		}
+		_file->seekg(_prevPos);
+	}
+	return type;
+}
+
+bool Parse::hasMoreToken()
+{
+	return !_file->eof();
+}
+
+Parse::~Parse()
+{
+
+}
+
+Json *Parse::parseObject(Json *node)
+{
+//	std::cout << "parsing object" << std::endl;
+	bool		endObj;
+
+	endObj = false;
+	while (!endObj)
+	{
+		if (hasMoreToken())
+		{
+			Type type;
+			type = getToken();
+			std::string value = type.value;
+			if (getToken().type != COLON)
+			{
+				std::cout << "bus error" << std::endl;
+				exit(0);
+			}
+			type = getToken();
+			switch (type.type)
+			{
+				case CURLY_OPEN:
+				{
+					node->type = Json::OBJECT;
+					parseTest(node);
+					_file->seekg(_prevPos);
+					break;
+				}
+				case CURLY_CLOSE:
+				{
+					endObj = true;
+					break;
+				}
+				case STRING:
+				{
+					node->type = Json::STRING;
+					node->values.str = type.value;
+					node->values.object.emplace(value, node);
+					break;
+				}
+				case NUMBER:
+				{
+					node->type = Json::NUMBER;
+					node->values.number = std::stoi(type.value);
+					node->values.object.emplace(value, node);
+					break;
+				}
+				case BOOLEAN:
+				{
+					node->type = Json::BOOLEAN;
+					(type.value == "true") ? node->values.boolean = true : node->values.boolean = false;
+					node->values.object.emplace(value, node);
+					break;
+				}
+				case COLON:
+					break;
+				case ARRAY_OPEN:
```

Comment:
For this case it would make sense to only call `parseArray` once if you encounter an array like this, no? I see it being called in a loop and that doesn't make much sense to me.

Reply:
I called in a loop as I can have nested array.
`[
      [
           [
            ]
        ]
   ]
` i can have this situation